### ISSUE TITLE
Tags and Grouping

### ISSUE THREAD

**Issue by smwurster** [springfox/springfox#1257]
The `tags` field on `@ApiOperation` says that it will override any value present at the `@Api` level.  Rather than being an override, this appears to be additive.  If I have the following:

    @Api(tags="Inventory")
    public class Foo {
        @ApiOperation(tags="Analysis")
        @RequestMapping(...)
        public String bar (...) { ... }
    }

The generated JSON will show that `bar()` has both the "Inventory" and "Analysis" tags, even though I only want it to contain the "Analysis" tag.  Is that a defect, or is that expected?

Related to this, is there a way for me to prevent the "description" field from being generated for a tag?  If you do not define a tag or description under `@Api`, the tag name in the JSON will default to the class name separated with dashes, and the description will be the class name separated by spaces.  How do I make it so the description field is never generated (I only want to use tags)?

I am using version 2.4.0.

Thanks.

**Comment by dilipkrish** [comment]
@smwurster the short answer is that it is expected. The reason being that the way these annotations are used in springfox is very different from the way swagger-core uses it. For e.g. we don't have support for constructs like `@SwaggerDefinition`. 

Regarding your second question, you can supply your [tags and descriptions](http://springfox.github.io/springfox/docs/current/#springfox-swagger2-with-spring-mvc-and-spring-boot) see 25. and it will not try and infer things.

**Comment by smwurster** [comment]
Okay, I can get the tags to come out as I want (i.e. without descriptions).  And I can stop the auto-generation of tags of the "class-name" style, as long as every use of the `Api` annotation defines a `tag` but not a `value`.

I still don't like the additive property of `tag` however, because it means you can't specify a tag at the `Api` level and then have routines in that class that define a different tag.  Those routines will end up with two tags instead of one.  Is there anyway I can solve this?

**Comment by dilipkrish** [comment]
Not at the moment. It seems weird that a controller would have operations on it that don't fall under the categorization of the controller. The work around I'd say is to provide different dockets for those methods.

**Comment by smwurster** [comment]
Some of our controllers cover a large general concept.  Let's take the example of a thing called `Item`.  Most of the operations in this controller are general metadata retrieval queries, and some are more complex queries.  They'll fall under an "Analysis" category.  But there are a few queries in this controller that are for downloading items; those fall under an "Export" category.  I don't want the Export-specific endpoints to show up under the Analysis category.  I'd rather not have to move functions into new controllers, as that will be time consuming (and annoying), and probably a little difficult to maintain in the future.

We used to solve this problem with a custom resource grouping strategy that was based on an annotation we defined.  The grouping strategy concept is not supported in springfox.  I did wonder whether I could dynamically define the tag at each function level via reflection, based on those annotations, but I didn't look into that.

Doesn't adding new dockets require users to select the proper group on the web page, since each docket defines its own group name?  We definitely don't want that.

**Comment by dilipkrish** [comment]
In that case my recommendation would be to implement your own [operation plugin](https://github.com/springfox/springfox/blob/master/springfox-spring-web/src/main/java/springfox/documentation/spring/web/readers/operation/OperationTagsReader.java#L42) that overwrites the values as per your need.

**Comment by smwurster** [comment]
Ah, I see.  At one point I tried creating an `ApiListingBuilderPlugin`, but I think it still ended up running the built-in one alongside mine.  I didn't look at the `OperationBuilderPlugin` but I will do that next.  I'll let you know how I make out.  Thanks.

**Comment by smwurster** [comment]
I created an `OperationBuilderPlugin`, and used it to find the annotations I wanted.  In that respect, it could work just like our resource grouping of old.

However, Springfox is still running its built-in equivalent, which means I'm getting tagging behaviors that I don't want.  Is there a way I can get it to stop using its built-in equivalent and just run mine?

**Comment by dilipkrish** [comment]
You need to set its priority (bean ordering) higher than the swagger plugins

**Comment by smwurster** [comment]
It looks like `OperationTagsReader` is set to highest precedence already, and I think `SwaggerOperationsTagReader` is as well.

**Comment by dilipkrish** [comment]
Are you [sure](https://github.com/springfox/springfox/blob/master/springfox-swagger-common/src/main/java/springfox/documentation/swagger/common/SwaggerPluginSupport.java#L30)? Setting it to anything higher than `(Ordered.HIGHEST_PRECEDENCE + 1000)` would just work right?

**Comment by smwurster** [comment]
Silly me, I was setting it relative to Spring's highest precedence, and not the one from `SwaggerOperationsTagReader`.  At the same time, I had commented out my setting of the all allowable tags in my `Docket`.  I fixed both of those and the tags are working great now.  Thanks.

Does my overriding of this affect any other built-in behavior?  Like is the hidden option on `ApiModelProperty` not processed, or anything like that?
